# Data cleaning using pandas

In this notebook, you'll see how clean structured data using `pandas`. We are aware that you have already done an extensive review of pandas, and are now experts in data manipulation, so we'll keep it short.

We are going to use a dataset about New York City Squirrels taken from [NYC Open Data](https://opendata.cityofnewyork.us/) which is a great resource for data sets.
This dataset is quite small and this notebook will just be an overview of different stuff to look out for when cleaning data.


In [1]:
import pandas as pd
import numpy as np

In [12]:
squirrel_data = r"C:\Users\pieta\OneDrive\Bureau\2018_Central_Park_Squirrel_Census_-_Squirrel_Data_20240715.csv"
df = pd.read_csv(squirrel_data)
df.head(5)

X          Y Unique Squirrel ID Hectare Shift      Date  \
0 -73.956134  40.794082     37F-PM-1014-03     37F    PM  10142018   
1 -73.968857  40.783783     21B-AM-1019-04     21B    AM  10192018   
2 -73.974281  40.775534     11B-PM-1014-08     11B    PM  10142018   
3 -73.959641  40.790313     32E-PM-1017-14     32E    PM  10172018   
4 -73.970268  40.776213     13E-AM-1017-05     13E    AM  10172018   

   Hectare Squirrel Number    Age Primary Fur Color Highlight Fur Color  ...  \
0                        3    NaN               NaN                 NaN  ...   
1                        4    NaN               NaN                 NaN  ...   
2                        8    NaN              Gray                 NaN  ...   
3                       14  Adult              Gray                 NaN  ...   
4                        5  Adult              Gray            Cinnamon  ...   

    Kuks  Quaas  Moans Tail flags Tail twitches  Approaches  Indifferent  \
0  False  False  False      False         False       False        False   
1  False  False  False      False         False       False        False   
2  False  False  False      False         False       False        False   
3  False  False  False      False         False       False        False   
4  False  False  False      False         False       False        False   

   Runs from  Other Interactions                                    Lat/Long  
0      False                 NaN  POINT (-73.9561344937861 40.7940823884086)  
1      False                 NaN  POINT (-73.9688574691102 40.7837825208444)  
2      False                 NaN  POINT (-73.97428114848522 40.775533619083)  
3       True                 NaN  POINT (-73.9596413903948 40.7903128889029)  
4      False                 NaN  POINT (-73.9702676472613 40.7762126854894)  

[5 rows x 31 columns]

In [13]:
df.describe()

X            Y          Date  Hectare Squirrel Number
count  3023.000000  3023.000000  3.023000e+03              3023.000000
mean    -73.967184    40.780853  1.011949e+07                 4.123718
std       0.007726     0.010285  4.246671e+04                 3.096492
min     -73.981159    40.764911  1.006202e+07                 1.000000
25%     -73.973102    40.771676  1.008202e+07                 2.000000
50%     -73.968594    40.778166  1.012202e+07                 3.000000
75%     -73.960189    40.791219  1.014202e+07                 6.000000
max     -73.949722    40.800119  1.020202e+07                23.000000

## Things to check

### Missing values

There are different ways that a value can be missing. It can be just an empty string (`""`), or `NaN` (Not a Number).
`np.nan`, `None` and `NaT` (for `datetime64` types) are standard missing value for Pandas.

You can either delete the rows/attribute if there are too many missing, or replace them. I can't give you a certain threshold to delete or replace, it really depends on the variable, on the context, and what you are going to do with the data later on.
In the same vein, there are many different ways to replace a value, and it will depend on the type of data you have. [[More info here]](https://pandas.pydata.org/pandas-docs/stable/user_guide/missing_data.html)

Pandas provides `isnull()`, `isna()` functions to detect missing values. Both of them do the same thing.

In [14]:
print("There are {} rows of data".format(len(df)))

There are 3023 rows of data


We see that we have 3023 rows of data. Looking at the number of NaN values per column, one can confidently say that columns where we have > 1000 missing values are not worth keeping. The case can be made that, for certain scenarios, the columns `Highlight Fur Color` should be kept too, but here we are going to drop it because the column `Combination of Primary and Highlight Color` combines both.

This notebook is going to feature small code blocks that you'll have to complete.

The first task is to count how many nan values there are per attribute. Hint: use `df.isnull()`

In [16]:
# Check how many rows of each attribute are NaN
df.isnull()


X      Y  Unique Squirrel ID  Hectare  Shift   Date  \
0     False  False               False    False  False  False   
1     False  False               False    False  False  False   
2     False  False               False    False  False  False   
3     False  False               False    False  False  False   
4     False  False               False    False  False  False   
...     ...    ...                 ...      ...    ...    ...   
3018  False  False               False    False  False  False   
3019  False  False               False    False  False  False   
3020  False  False               False    False  False  False   
3021  False  False               False    False  False  False   
3022  False  False               False    False  False  False   

      Hectare Squirrel Number    Age  Primary Fur Color  Highlight Fur Color  \
0                       False   True               True                 True   
1                       False   True               True                 True   
2                       False   True              False                 True   
3                       False  False              False                 True   
4                       False  False              False                False   
...                       ...    ...                ...                  ...   
3018                    False  False              False                 True   
3019                    False  False              False                False   
3020                    False  False              False                False   
3021                    False   True              False                False   
3022                    False  False              False                False   

      ...   Kuks  Quaas  Moans  Tail flags  Tail twitches  Approaches  \
0     ...  False  False  False       False          False       False   
1     ...  False  False  False       False          False       False   
2     ...  False  False  False       False          False       False   
3     ...  False  False  False       False          False       False   
4     ...  False  False  False       False          False       False   
...   ...    ...    ...    ...         ...            ...         ...   
3018  ...  False  False  False       False          False       False   
3019  ...  False  False  False       False          False       False   
3020  ...  False  False  False       False          False       False   
3021  ...  False  False  False       False          False       False   
3022  ...  False  False  False       False          False       False   

      Indifferent  Runs from  Other Interactions  Lat/Long  
0           False      False                True     False  
1           False      False                True     False  
2           False      False                True     False  
3           False      False                True     False  
4           False      False                True     False  
...           ...        ...                 ...       ...  
3018        False      False                True     False  
3019        False      False                True     False  
3020        False      False                True     False  
3021        False      False                True     False  
3022        False      False                True     False  

[3023 rows x 31 columns]

Now that we've identified all the columns that have NaN values, we would like to drop those with more than 1000 NaN values.

Can you do that? Hint: `df.drop()`
  
Moreover, can you print the columns that were dropped, along with the number of missing values that they had?

In [19]:
# Drop the columns that have more than 1000 missing values


        
# Print the columns that were dropped


TypeError: '>' not supported between instances of 'list' and 'int'

### Attribute type

Sometimes, an attribute does not have the correct type, e.g. integers are labeled as floats. So it is useful to check if it needs to be changed or not.


Please print the dtypes of every attribute.

In [ ]:
# Print the dtype of every attribute


For example, for `Hectare`, one could think that the variable is just the size of the plot of land. However, it seems it is just an identifier for which plot of land it is. One would have to see the documentation to know for sure. In this case, the `object` dtype is appropriate.

In [ ]:
# Print unique values of the "Hectare" attribute

### Attribute values

Carefully check each attribute to see if the unique values are coherent and make sense.
Moreover, check that minimum and maximum values make sense. (e.g. age of a person that is negative)

Hint: use your previously acquired knowledge of extracting unique values.
Be careful to limit the attributes that you print by the number of unique values they have, as there are fields that have all their values that are unique, and we don't want a kilometer long list of each value.

In [ ]:
# Print the number of unique values of each column, only printing those that have less than 100 unique values


In [ ]:
df.describe(include="all").T  # Transpose the data frame so that it fits in a cell

**Does something stand out to you?**

In the column `Combination of Primary and Highlight Color` for example, we have a value that is only `"+"`.
Again, there are multiple ways to deal with this. When you can't figure out what the value means (like in this case), you can drop it. Sometimes, when you are able to figure out which one is meant, you can replace it with the correct value. This is where the domain knowledge come into play, knowing which variable can be replaced with what, and what it could mean.

Another example, from another dataset containing vehicle types where the vehicle type was manually encoded.
You see how `"Tow truck"` appears in many different forms.
The right approach is to consolidate all of the different values of tow truck into a single one.

In [ ]:
trucks = ["Pick-up Truck", "Tractor Truck Diesel", "Box Truck", "Tractor Truck Gasoline", "Tow Truck / Wrecker", "Beverage Truck", "truck", "Armored Truck", "TRUCK", "Truck", "Tow Truck", "FIRE TRUCK", "PICK-UP TRUCK", "BOX TRUCK", "Fire Truck", "box truck", "TRUCK FLAT", "FIRTRUCK", "TRUCK VAN", "FIRETRUCK", "FDNY TRUCK", "dump truck", "Fire truck", "UPS TRUCK", "fire truck", "FOOD TRUCK", "MAIL TRUCK", "USPS TRUCK", "tow truck", "TOW TRUCK", "Tow truck"]
[truck for truck in trucks if "tow" in truck.lower()]

### Duplicates

Search for duplicates and remove them



In [ ]:
# There are no duplicates
df.duplicated().any()

If there had been, call `df.drop_duplicates(inplace=True)` to remove them.

Again, this is just a short introduction to data cleaning of structured data, there are many more steps that one could optionally take.

## References and reading material

[Data cleaning challenge - Kaggle](https://www.kaggle.com/rtatman/data-cleaning-challenge-handling-missing-values)  
[Preparing data for a machine learning model - Dataquest](https://www.dataquest.io/blog/machine-learning-preparing-data/)  
[Preliminary data cleaning - Kaggle](https://www.kaggle.com/quannguyen135/preliminary-data-cleaning-with-python)  
[Pandas data cleaning cheatsheet - Elite Data Science](https://elitedatascience.com/python-cheat-sheet)